build/parseSpec.c

```diff
@@ -171,20 +171,64 @@ static int restoreFirstChar(rpmSpec spec)
     return 0;
 }
 
+static const struct lineTypes {
+    int id;
+    size_t textLen;
+    const char *text;
+    int withArgs;
+} specLineTypes[] = {
+    { LINE_IF,          LEN_AND_STR("%if")     , 1},
+    { LINE_IFARCH,      LEN_AND_STR("%ifarch") , 1},
+    { LINE_IFNARCH,     LEN_AND_STR("%ifnarch"), 1},
+    { LINE_IFOS,        LEN_AND_STR("%ifos")   , 1},
+    { LINE_IFNOS,       LEN_AND_STR("%ifnos")  , 1},
+    { LINE_ELSE,        LEN_AND_STR("%else")   , 0},
+    { LINE_ENDIF,       LEN_AND_STR("%endif")  , 0},
+    { LINE_INCLUDE,     LEN_AND_STR("%include"), 1},
+    { LINE_OTHER,       0, NULL                }
+};
+static const size_t numLineTypes = sizeof(specLineTypes)/sizeof(*specLineTypes);
+
+static int parseLineType(char *line)
+{
+    for (int i=0; i<(numLineTypes-1); i++) {
+	if (specLineTypes[i].withArgs) {
+	    if ( ISMACROWITHARG(line, specLineTypes[i].text, specLineTypes[i].textLen) )
+		return i;
+	} else {
+	    if ( ISMACRO(line, specLineTypes[i].text, specLineTypes[i].textLen) )
+		return i;
+	}
+    }
+
+    return numLineTypes-1;
+}
+
 static int expandMacrosInSpecBuf(rpmSpec spec, int strip)
 {
     char *lbuf = NULL;
+    char *auxBuf;
     int isComment = 0;
-
-     /* Don't expand macros (eg. %define) in false branch of %if clause */
-    if (!spec->readStack->reading)
-	return 0;
+    int lineTypeNumber;
 
     lbuf = spec->lbuf;
     SKIPSPACE(lbuf);
     if (lbuf[0] == '#')
 	isComment = 1;
 
+    lineTypeNumber = parseLineType(lbuf);
+    if (specLineTypes[lineTypeNumber].id & (LINE_ELSE | LINE_ENDIF)) {
```

Comment:
The more appropriate test here would be whether the current directive type takes an argument or not.